### PR TITLE
feat(inference): add support CRUD deployment

### PR DESCRIPTION
**Pull request by Laure-di**

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package provider
 
 import (
 	"context"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/services/inference"
```
Comment (remyleone) [inline internal/provider/provider.go:5]: Please make this import align nicely with the other using gofumpt

File: internal/services/inference/helpers_inference.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package inference
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	inference "github.com/scaleway/scaleway-sdk-go/api/inference/v1beta1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/locality/regional"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/meta"
+	"time"
```
Comment (remyleone) [inline internal/services/inference/helpers_inference.go:9]: Need to be gofumpt

File: internal/services/inference/deployment.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+	inference "github.com/scaleway/scaleway-sdk-go/api/inference/v1beta1"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/httperrors"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/locality/regional"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/services/account"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/types"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+	_ "time"
```
Comment (remyleone) [inline internal/services/inference/deployment.go:14]: Why is this import ignored?

File: internal/services/inference/deployment.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+import (
+	"context"
+	inference "github.com/scaleway/scaleway-sdk-go/api/inference/v1beta1"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/httperrors"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/locality/regional"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/services/account"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/types"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
```
Comment (remyleone) [inline internal/services/inference/deployment.go:11]: gofumpt

File: internal/services/inference/deployment.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+	if maxSize, ok := d.GetOk("max_size"); ok {
+		req.MaxSize = scw.Uint32Ptr(uint32(maxSize.(int)))
+	}
+
+	if minSize, ok := d.GetOk("min_size"); ok {
+		req.MaxSize = scw.Uint32Ptr(uint32(minSize.(int)))
+	}
+
+	req.Endpoints = []*inference.EndpointSpec{&endpoint}
```
Comment (remyleone) [inline internal/services/inference/deployment.go:169]: Move the code that is referring to endpoints closer together to reduce cognitive load

File: internal/services/inference/deployment_test.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/acctest"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/httperrors"
+	"github.com/scaleway/terraform-provider-scaleway/v2/internal/services/inference"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccDeployment_Basic(t *testing.T) {
```
Comment (remyleone) [inline internal/services/inference/deployment_test.go:15]: Could you add a test with a private endpoint?

File: internal/services/inference/deployment.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+						"public_endpoint": {
+							Type:        schema.TypeBool,
+							Description: "Set the endpoint as public",
+							Optional:    true,
+						},
+						"private_endpoint": {
+							Type:        schema.TypeString,
+							Description: "The id of the private network",
+							Optional:    true,
+						},
```
Comment (Codelax) [inline internal/services/inference/deployment.go:146]: Having `public_endpoint` as a bool and `private_endpoint` as an ID is misleading.
These fields should have names that suit their use, example: `is_public`, `private_network_id`

File: internal/services/inference/deployment.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+	_ = d.Set("model_name", deployment.ModelName)
+	_ = d.Set("min_size", int(deployment.MinSize))
+	_ = d.Set("max_size", int(deployment.MaxSize))
+	_ = d.Set("size", int(deployment.Size))
+	_ = d.Set("status", deployment.Status)
+	_ = d.Set("model_id", deployment.ModelID)
+	_ = d.Set("created_at", types.FlattenTime(deployment.CreatedAt))
+	_ = d.Set("updated_at", types.FlattenTime(deployment.UpdatedAt))
+
+	for _, endpoint := range deployment.Endpoints {
```
Comment (Codelax) [inline internal/services/inference/deployment.go:241]: Should this export as `endpoints` ? User configure its `endpoints` but get different fields after StateRead.

File: internal/services/inference/deployment.go [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+				Description: "The endpoint public ID",
+			},
+			"endpoint_private_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The endpoint private ID",
+			},
+			"disable_auth_public": {
+				Type:        schema.TypeBool,
+				Computed:    true,
```
Comment (remyleone) [inline internal/services/inference/deployment.go:118]: Should an optional be present here?

File: internal/services/inference/deployment.go [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+				Description: "The endpoint public URL",
+			},
+			"endpoint_private_url": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The endpoint private URL",
+			},
+			"disable_auth_private": {
+				Type:        schema.TypeBool,
+				Computed:    true,
```
Comment (remyleone) [inline internal/services/inference/deployment.go:103]: Should an optional be present here?

File: internal/services/inference/deployment.go (hunk 8 above)
Comment (remyleone) [reply internal/services/inference/deployment.go:241]: It doesn't make much sense to have a field at the root of the ressource being updated with a sub resource downstream :/ would it be better to have all information about endpoint at the same place?

File: internal/services/inference/deployment.go [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "The id of the public endpoint",
+							Computed:    true,
+						},
+						"is_enabled": {
+							Type:        schema.TypeBool,
+							Description: "Set the endpoint as public",
```
Comment (Codelax) [inline internal/services/inference/deployment.go:150]:
```suggestion
							Description: "Enable or disable public endpoint",
```

File: internal/services/inference/deployment.go [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+				AtLeastOneOf: []string{"public_endpoint"},
+				Description:  "List of endpoints",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "The id of the private endpoint",
+							Computed:    true,
+						},
+						"private_endpoint_id": {
```
Comment (Codelax) [inline internal/services/inference/deployment.go:115]:
```suggestion
						"private_network_id": {
```
You will also need to change in the rest of the code.

File: internal/services/inference/deployment.go [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+			publicEndpoints = append(publicEndpoints, publicEndpointSpec)
+		}
+	}
+
+	if len(privateEndpoints) > 0 {
+		_ = d.Set("private_endpoint", privateEndpoints)
+	}
+	if len(publicEndpoints) > 0 {
+		_ = d.Set("public_endpoint", publicEndpoints)
+	}
```
Comment (Codelax) [inline internal/services/inference/deployment.go:292]: MaxItems is 1 for `private_endpoint` and `public_endpoint`. This code make it seems multiple endpoints are supported. Should we let user specify multiple endpoints in terraform ? If API allows it, terraform provider would be ready.

File: internal/services/inference/deployment.go [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+		return diag.FromErr(err)
+	}
+	req := &inference.UpdateDeploymentRequest{
+		Region:       region,
+		DeploymentID: deployment.ID,
+	}
+
+	if d.HasChange("name") {
+		req.Name = types.ExpandUpdatedStringPtr(d.Get("name"))
+	}
```
Comment (Codelax) [inline internal/services/inference/deployment.go:317]: Api allows more field in Update: https://github.com/scaleway/scaleway-sdk-go/blob/a2c1e85de57a916ea17d27c4efe2646fc5f33d11/api/inference/v1beta1/inference_sdk.go#L786-L803

File: internal/services/inference/deployment.go [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Optional:    true,
+				Description: "The tags associated with the deployment",
+			},
+			"min_size": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+				Description: "The minimum size of the pool",
```
Comment (remyleone) [inline internal/services/inference/deployment.go:73]: Should add validation about the int being positive. Can the `min_size` be 0?

File: internal/services/inference/deployment.go (hunk 13 above)
Comment (Laure-di) [reply internal/services/inference/deployment.go:292]: You can have only one private network and only one public

File: internal/services/inference/deployment.go [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+							Type:        schema.TypeString,
+							Description: "The id of the public endpoint",
+							Computed:    true,
+						},
+						"is_enabled": {
+							Type:        schema.TypeBool,
+							Description: "Set the endpoint as public",
+							Optional:    true,
+						},
+						"disable_auth": {
```
[the remark this answers is not in the document]
Comment (Laure-di) [reply internal/services/inference/deployment.go:153]: by default, deployments are protected by IAM authentication.

So by default the disable_auth is false if not specified

File: internal/services/inference/deployment.go (hunk 15 above)
Comment (remyleone) [reply internal/services/inference/deployment.go:73]: I would prefer it to maintain the computed

**Pull request merged**